### PR TITLE
Calculate mean done.

### PR DESCRIPTION
**Pull request by TekaMesfinAbel**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/calculate_mean.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+"""
+This module provides a function to calculate the mean (average) of a list of numbers.
+
+Author: @Abel Teka
+"""
```
Comment (ViktoriyaHaiduk) [inline solutions/calculate_mean.py:5]: Great work, Abel. Due to Even's example, let's put date.

File: solutions/calculate_mean.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    Calculate the mean (average) of a list of numbers.
+
+    Args:
+        numbers (List[float]): A list of numbers (integers or floats).
+            - The list can include positive, negative, or zero values.
+
+    Returns:
+        Optional[float]: The mean (average) of the numbers in the list.
+            - Returns None if the list is empty.
```
Comment (ViktoriyaHaiduk) [inline solutions/calculate_mean.py:20]: It might be helpful to include an explanation of what happens if the list contains only one item. This would provide clarity on how the functionality works in such cases

File: solutions/tests/test_calculate_mean.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        """Test that an empty list returns None."""
+        self.assertIsNone(calculate_mean([]))
+
+    def test_boundary_case_zero_values(self):
+        """Test mean calculation with a list of zeros."""
+        self.assertEqual(calculate_mean([0, 0, 0, 0]), 0.0)
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (ViktoriyaHaiduk) [inline solutions/tests/test_calculate_mean.py:88]: Everything looks great! Well done! I really appreciate the comprehensive set of unit tests.

**Review by ViktoriyaHaiduk — DISMISSED** [review]
Excellent job! Your thorough approach and attention to detail are impressive. The comprehensive set of unit tests reflects your dedication and expertise

File: solutions/calculate_mean.py (hunk 1 above)
Comment (TekaMesfinAbel) [reply solutions/calculate_mean.py:5]: Thanks, Viktoriya, okay will do that.

File: solutions/calculate_mean.py (hunk 2 above)
Comment (TekaMesfinAbel) [reply solutions/calculate_mean.py:20]: Thank you for the suggestion! I see how adding an explanation for cases where the list contains only one item could improve clarity. In the current implementation, if the list has a single item, the function will simply return that item as the mean, since the mean of a single number is the number itself. I'll update the docstring to include this clarification.

File: solutions/tests/test_calculate_mean.py (hunk 3 above)
Comment (TekaMesfinAbel) [reply solutions/tests/test_calculate_mean.py:88]: Thank you.

File: solutions/tests/test_calculate_mean.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            calculate_mean("not a list")
+
+    def test_invalid_input_non_numeric_elements(self):
+        """Test for TypeError when the list contains non-numeric elements."""
+        with self.assertRaises(TypeError):
+            calculate_mean([1, 2, "three", 4])
+
+    def test_invalid_input_nan(self):
+        """Test for ValueError when the list contains NaN."""
+        with self.assertRaises(ValueError):
```
Comment (GaiSamuel) [inline solutions/tests/test_calculate_mean.py:63]: The unittest doesn't raise a value error. it's best to leave it as self.assertRaises(Exception) because it neither raises type error nor assertion error. Adjust the docstrings accordingly.

File: solutions/tests/test_calculate_mean.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            calculate_mean([1, 2, float("nan"), 4])
+
+    def test_invalid_input_infinity(self):
+        """Test for ValueError when the list contains infinity."""
+        with self.assertRaises(ValueError):
```
Comment (GaiSamuel) [inline solutions/tests/test_calculate_mean.py:68]: changes same as Line 63 above

File: solutions/calculate_mean.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        5.0
+        >>> calculate_mean([1.5, 2.5, 3.5])
+        2.5
+    """
+    if not isinstance(numbers, list):
+        raise TypeError("Input must be a list.")
+    if any(not isinstance(n, (int, float)) for n in numbers):
+        raise TypeError("All elements in the list must be numbers.")
+    if any(x != x or x in {float("inf"), float("-inf")} for x in numbers):
+        raise ValueError(
```
Comment (GaiSamuel) [inline solutions/calculate_mean.py:44]: adjust this according to the changes made in the tests in line 63 and 68.

File: solutions/calculate_mean.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        numbers (List[float]): A list of numbers (integers or floats).
+            - The list can include positive, negative, or zero values.
+
+    Returns:
+        Optional[float]: The mean (average) of the numbers in the list.
+            - Returns None if the list is empty.
+
+    Raises:
+        TypeError: If `numbers` is not a list or contains non-numeric elements.
+        ValueError: If the list contains infinite or NaN values.
```
Comment (GaiSamuel) [inline solutions/calculate_mean.py:25]: adjust this as well according to the changes suggested in the tests' lines 63 and 68

**Review by GaiSamuel — CHANGES_REQUESTED** [review]
Finished with my review. Great work done! I just suggested a few changes.

File: solutions/tests/test_calculate_mean.py (hunk 4 above)
Comment (TekaMesfinAbel) [reply solutions/tests/test_calculate_mean.py:63]: ohh I didn't catch that. thank you.

**Review by ViktoriyaHaiduk — APPROVED** [review]
Good job!

**Review by ViktoriyaHaiduk — APPROVED** [review]
Good job! Everything is according to check list.

**Review by GaiSamuel — APPROVED** [review]
Looks good! Well, done Abel.

**Review by ViktoriyaHaiduk — APPROVED** [review]
Great job on the implementation! Your function is well-structured, and the documentation clearly explains how it works. Overall, the code is clean, concise, and easy to understand. Keep up the good work!
